com.unity.multiplayer.mlapi/Runtime/Core/RPCQueue/RPCQueueHistoryFrame.cs

```diff
@@ -0,0 +1,226 @@
+
+            //Get the stream size
+            CurrentQueueItem.StreamSize = QueueReader.ReadInt64();
+
+            //Inbound and Outbound message streams are handled differently
+            if (queueFrameType == QueueFrameType.Inbound)
+            {
+                //Get our offset
+                long Position = QueueReader.ReadInt64();
+                //Always make sure we are positioned at the start of the stream before we write
+                CurrentQueueItem.ItemStream.Position = 0;
+
+                //Write the entire message to the CurrentQueueItem stream (1 stream is re-used for all incoming RPCs)
+                CurrentQueueItem.StreamWriter.ReadAndWrite(QueueReader, CurrentQueueItem.StreamSize);
+
+                //Reset the position back to the offset so std rpc API can process the message properly
+                //(i.e. minus the already processed header)
+                CurrentQueueItem.ItemStream.Position = Position;
+            }
+            else
+            {
+                //Create a byte array segment for outbound sending
+                CurrentQueueItem.MessageData = QueueReader.CreateArraySegment((int)CurrentQueueItem.StreamSize, (int)QueueStream.Position);
+            }
+
+            return CurrentQueueItem;
+        }
+
+        /// <summary>
+        /// GetNextQueueItem
+        /// Handles getting the next queue item from this frame
+        /// If none are remaining, then it returns a queue item type of NONE
+        /// </summary>
+        /// <returns>FrameQueueItem</returns>
+        public FrameQueueItem GetNextQueueItem()
+        {
+            QueueStream.Position = QueueItemOffsets[QueueItemOffsetIndex];
+            QueueItemOffsetIndex++;
+            if (QueueItemOffsetIndex >= QueueItemOffsets.Count)
+            {
+                CurrentQueueItem.QueueItemType = RPCQueueManager.QueueItemType.None;
+                return CurrentQueueItem;
+            }
+
+            return GetCurrentQueueItem();
+        }
+
+        /// <summary>
+        /// GetFirstQueueItem
+        /// Should be called the first time a queue item is pulled from a queue history frame.
+        /// This will reset the frame's stream indices and add a new stream and stream writer to the CurrentQueueItem instance.
+        /// </summary>
+        /// <returns>FrameQueueItem</returns>
+        public FrameQueueItem GetFirstQueueItem()
+        {
+            if (QueueStream.Position > 0)
+            {
+                QueueItemOffsetIndex = 0;
+                QueueStream.Position = 0;
+
+                if (queueFrameType == QueueFrameType.Inbound)
+                {
+                    CurrentQueueItem.ItemStream = PooledBitStream.Get();
+                    CurrentQueueItem.StreamWriter = PooledBitWriter.Get(CurrentQueueItem.ItemStream);
+                    CurrentQueueItem.StreamReader = PooledBitReader.Get(CurrentQueueItem.ItemStream);
+                }
+
+                return GetCurrentQueueItem();
```

Comment:
is this an error condition? (if not, should we comment inline?)

Reply:
This is not an error condition, just no entries left to get.